NachoSpigot-Server/src/main/java/dev/cobblesword/nachospigot/hitdetection/LagCompensator.java

```diff
@@ -0,0 +1,79 @@
+package dev.cobblesword.nachospigot.hitdetection;
+
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.ListMultimap;
+import me.elier.nachospigot.config.NachoConfig;
+import org.apache.commons.lang3.tuple.Pair;
+import org.bukkit.Location;
+import org.bukkit.entity.Player;
+import org.bukkit.util.Vector;
+
+import java.util.*;
+
+public class LagCompensator {
+
+    /*
+        Lag Compensator based on https://github.com/Islandscout/ServerSideHitDetection
+        Adjusts hit detection to be more fair for players with worse ping.
+     */
+
+    private final ListMultimap<UUID, Pair<Location, Long>> locationTimes = ArrayListMultimap.create();
+    private final int historySize = 40;
+    private final int pingOffset = 92;
+    private final int timeResolution = 30;
+
+    public Location getHistoryLocation(Player player) {
+        if (!locationTimes.containsKey(player.getUniqueId()))
+            return player.getLocation();
+
+        int rewindMillisecs = player.spigot().getPing();
+
+        List<Pair<Location, Long>> times = locationTimes.get(player.getUniqueId());
```

Comment:
Personally, I would probably make this a class instead of a pair, I would probably call it LocationSnapshot, and state the double x, y, z; long registeredTimestamp; so we don't hold that heavy world reference by mistake somehow.

Also for the variable name, I would probably call it "previousLocations"